File: react/aaudiocore/src/models/play-command.ts
import type { IPlaybackStrategy } from "../contracts/playback-strategy";
import type { PlaybackTrack } from "./track";

export enum PlayCommandType {
  Play = "play",
  Pause = "pause",
  Stop = "stop",
  Toggle = "toggle",
  Next = "next",
  Previous = "previous",
  SetVolume = "setVolume",
  SetPlaybackRate = "setPlaybackRate",
  SetPlaybackStrategy = "setPlaybackStrategy",
  SetQueue = "setQueue"
}

export type IPlayCommand =
  | { type: PlayCommandType.Play }
  | { type: PlayCommandType.Pause }
  | { type: PlayCommandType.Stop }
  | { type: PlayCommandType.Toggle }
  | { type: PlayCommandType.Next }
  | { type: PlayCommandType.Previous }
  | { type: PlayCommandType.SetVolume; volume: number }
  | { type: PlayCommandType.SetPlaybackRate; playbackRate: number }
  | { type: PlayCommandType.SetPlaybackStrategy; strategy: IPlaybackStrategy }
  | { type: PlayCommandType.SetQueue; queue: PlaybackTrack[]; startIndex?: number };

File: winforms/aaudiocore/Models/PlayCommand.cs
using AAudioCore.Contracts;

namespace AAudioCore.Models;

public enum PlayCommandType
{
    Play,
    Pause,
    Stop,
    Toggle,
    Next,
    Previous,
    SetVolume,
    SetPlaybackRate,
    SetPlaybackStrategy,
    SetQueue
}

public sealed record PlayCommand
{
    public PlayCommandType Type { get; init; }
    public float Volume { get; init; }
    public float PlaybackRate { get; init; }
    public IReadOnlyList<PlaybackTrack>? Queue { get; init; }
    public int StartIndex { get; init; }
    public IPlaybackStrategy? Strategy { get; init; }

    public static PlayCommand Create(PlayCommandType type) => new() { Type = type };
}
